Override API Gateway's values for API documentation (#1436)

jessebrennan opened the pull request:
Before this change, the specs were combined. But now, if any
documentation is provided, we will assume it is correct and complete.
Because of this, it should override the default values provided by API
Gateway.

danielsotirhos commented on lambdas/service/app.py:
```diff
@@ -184,13 +184,22 @@ def openapi():
+            'name': 'keys',
+            'in': 'path',
+            'required': True,
+            'schema': {'type': 'string'}
```
This is specifying that the `/health` endpoint has a required path parameter `keys`. This is incorrect, no?
`/health/{keys}` has a required path parameter but `/health` does not.

jessebrennan commented:
@hannes-ucsc, I used my discretion to redo the branches history. I did this because the new code is so drastically different than what was there before, the commit history would no longer have made sense (especially as this PR now solved what was #1454).

A few interesting points:
* Although in conversation we had agreed that we should have separate documentation for the `/health/` route, this isn't actually possible in practice. To document this path we would need to add a separate `@route` decorator, but Chalice doesn't allow trailing slashes on routes:
  ```
  Traceback (most recent call last):
    File "/home/jesse/gi/azul/.venv/lib/python3.6/site-packages/chalice/deploy/deployer.py", line 371,   in _validate_config
      validate_configuration(config)
    File "/home/jesse/gi/azul/.venv/lib/python3.6/site-packages/chalice/deploy/validate.py", line 41, in validate_configuration
      validate_routes(routes)
    File "/home/jesse/gi/azul/.venv/lib/python3.6/site-packages/chalice/deploy/validate.py", line 107, in validate_routes
      % route_name)
  ValueError: Route cannot end with a trailing slash: /health/
  ```
  Instead I included this information in the `/health/{keys}` path, but then Swagger doesn't allow you to then exercise the `/health/` path through the 'try it out functionality'
![required-health](https://user-images.githubusercontent.com/10734740/69299734-288d4b00-0bc6-11ea-8c8c-46ea5f248276.png)
  Unfortunately this is a limitation of having the API this way.
* When reviewing the code you'll probably notice that there is some spec that's shared between health endpoints. This is unavoidable given the limitations of OpenAPI. See the commit message for more details.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -2,23 +2,65 @@
 
 
-def openapi_spec(description: JSON):
+def openapi_spec(path: str, methods: List[str], path_spec: MutableJSON, method_spec: MutableJSON):
```
Shouldn't `path_spec` be optional?

hannes-ucsc replied:
In fact, I think both need to be optional. We should force users to pass these as keyword arguments. We should programmatically enforce that at least one of the two is passed.

jessebrennan replied:
I added this and tests.

jessebrennan replied:
But actually I have to remove this behaviour for now. If we prevent both from being None, then any existing endpoints without documentation will cause this assertion to fail. Once all endpoints are documented, we can consider adding this back in.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -2,23 +2,65 @@
+    ... @openapi_spec('/bar', ['GET'], {'a': 'b'}, {'c': 'd'})
+    ... def bar():
+    ...     pass
+    Traceback (most recent call last):
```
What about 

```
@openapi_spec('/bar', ['PUT'], {…}, {…})
def foo():
    pass
@openapi_spec('/bar', ['GET'], {…}, {…})
def bar():
    pass
```

Should that not raise an exception, too?

jessebrennan replied:
This is an important error case to check for, but it can't be caught here. It's needs to be tested for in `join_specs`.

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -184,20 +184,40 @@ def openapi():
-            'schema': {'$ref': "#/components/schemas/Empty"}
+            'name': 'keys',
+            'in': 'path',
+            'required': True,
```
So despite being required, it's still allowed to be empty? Does `GET /health/` still work?

jessebrennan replied:
Path parameters are always required https://swagger.io/docs/specification/describing-parameters/#path-parameters. You can still make the request, just not through the Swagger page.

jessebrennan commented on lambdas/service/app.py:
```diff
@@ -184,20 +184,40 @@ def openapi():
 })
-@app.route('/health/{keys}', methods=['GET'], cors=True)
-def health(keys: Optional[str] = None):
+def health_keys(keys: Optional[str] = None):
```
Woops! Thanks for catching this.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -37,32 +109,52 @@ def annotated_specs(gateway_id, app, toplevel_spec) -> JSON:
-    return docs
+        if hasattr(func, 'path_specs'):
+            for path, spec in func.path_specs.items():
+                if path in path_specs:
```
I noticed you do this here and there. It's not explicitly specified in the contribution guide but we follow the commonly accepted [EAFP](https://stackoverflow.com/questions/11360858/what-is-the-eafp-principle-in-python) idiom. If you feel it should be mentioned, could you add a section to the guidelines? Either way, though, this should use EAFP.

jessebrennan replied:
Sure. I find LBYL to be more legible (and often less code), but I understand the strengths of EAFP, especially with regard to race conditions. I'm happy to switch it over.

hannes-ucsc replied:
Despite EAFP being more verbose, it IS easier to reason about because it doesn't do everything twice.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -37,32 +109,52 @@ def annotated_specs(gateway_id, app, toplevel_spec) -> JSON:
+        # This may override duplicate specs from path_specs or API Gateway
+        specs['paths'][path][method] = method_spec
+
+    return specs
```
Looks like this code isn't covered by any tests: https://codecov.io/gh/DataBiosphere/azul/pull/1449/diff#D3-127.

jessebrennan replied:
It is covered by the integration test. If you want better coverage, I can add more tests.

hannes-ucsc replied:
*Some* of it is covered by IT. The error case isn't. Correct me if I am wrong but during my penultimate review the two-handlers-same-path case was not tested for and NOT handled. In response to that review you added a check for that to the code under test but the code is not present. So yes, I would like *unit* test coverage for at least the error case but ideally all of `get_app_specs`. 

Re `join_specs` not being covered by IT: it looks like it would be easy to do with a few doctests.

For future reference: when I say "coverage" or "covered" without qualification, I'm referring to the unit test coverage as measured by Coveralls and Codecov.

hannes-ucsc replied:
Sorry for the multiple edits. Done now.

jessebrennan replied:
Ok. I agree and I'll add this.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -125,25 +142,70 @@ def clean_specs(specs):
     for entries in app.routes.values():
         func = next(iter(entries.values())).view_function
-        if hasattr(func, 'path_specs'):
+        try:
```
Each `.` in the body of this try could potentially raise an AttributeError. The body of the `try` should be minimal (one `.` or one use of `[]` for dicts) for so as to not mask/conflate other issues.

Also note that it IS ok to use `dict.get()` and `getattr()` with a default since they use EAFP under the hood.

This might seem pedantic, but this is worse now than the previous version. I try to stick with EAFP as habit so I don't need to think about when it does matter (performance-critical or concurrent code) and when it doesn't.

jessebrennan replied:
I agree this is worse. Good point about `getattr()`. I've switched to that since I think it is the most readable, yet still EAFP.

hannes-ucsc commented on src/azul/openapi.py:
```diff
@@ -125,25 +142,70 @@ def clean_specs(specs):
-            path_specs.update(func.path_specs)
-        else:
+        except AttributeError:
             # TODO (jesse): warn if route is undocumented
```
Is this hard to do? When do you plan to do this?

jessebrennan replied:
This can't be done until all of the routes are documented.

hannes-ucsc replied:
You can't *warn* unless all routes are documented?

jessebrennan replied:
I changed "warn" to "raise", because that was what I intended, and it cannot be done until all of the routes are documented.

jessebrennan replied:
I don't think we want to _actually_ make a warning log here because then it will show up in Airbrake, etc. which I believe we don't want.

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -184,20 +184,40 @@ def openapi():
 })
-@app.route('/health/{keys}', methods=['GET'], cors=True)
-def health(keys: Optional[str] = None):
+def health_keys(keys: Optional[str] = None):
```
Old change that crept back in.

hannes-ucsc commented on scripts/envhook.py:
```diff
@@ -1,17 +1,19 @@
 #! /usr/bin/env python3
 
 import errno
+from importlib.abc import (
```
Why is this change in this PR?

jessebrennan replied:
I was experimenting with PyCharm's changelists, and had your envhook fix saved separately in order to run doctests. When I was rebasing I forgot to "shelve" the envhook fix so it got roped in.

jessebrennan commented:
@hannes-ucsc proposed changing the implementation to store the spec with the Chalice app object. This is a better design than the current solution, but for expediency, and since this current solution is basically ready to merge, I think this re-design should be addressed later, in https://github.com/DataBiosphere/azul/issues/1485.

jessebrennan commented:
@hannes-ucsc I see your point about the type signatures. I'll pay more attention next time when I compose them. I added (and approved!) your changes into a new fixup commit and rebased.
